autoPyTorch/pipeline/components/training/trainer/base_trainer.py

```diff
@@ -0,0 +1,351 @@
+import logging
+import time
+from typing import Any, Callable, Dict, List, Optional, Tuple
+
+import numpy as np
+
+import torch
+from torch.autograd import Variable
+from torch.utils.tensorboard.writer import SummaryWriter
+
+from autoPyTorch.pipeline.components.training.base_training import autoPyTorchTrainingComponent
+
+
+class BudgetTracker(object):
+    def __init__(self,
+                 budget_type: str,
+                 max_value: int,
+                 ):
+        """
+        An object for tracking when to stop the network training.
+        It handles epoch based criteria as well as training based criteria.
+
+
+        # Maybe it is better to use pynisher? Remember the whole everything gets
+        copied when forking memory, so it makes sense to ask... but also,
+        how to we regulate memory from GPU and CPU?
```

Comment:
I dont think it is necessary to regulate GPU/CPU memory or do you have a case where it would be? Typically you use all available cores / GPUs and those should be given by the Optimizer/Scheduler

Reply:
Yeah, I agree the scheduler should manage. This was just a reminder that the scheduler has to be gpu aware, because we don't want to run out of memory in the GPU because we over scheduled. This was a reminder that pynisher cannot control this today nor smac is aware of this (yet).